lib/flight_scheduler/event_processor.rb

```diff
@@ -93,7 +93,13 @@ def node_completed_job(node_name, job_id)
     return if job.nil?
 
     Async.logger.info("Node #{node_name} completed job #{job.display_id}")
-    job.state = 'COMPLETED'
+    if job.state == 'CANCELLING'
+      job.state = 'CANCELLED'
```

Comment:
I don't think this is right.  It's possible that a request is made to cancel the job, but before the daemon acts on that request the job completes.  In that case, I think the job has completed rather than cancelled.

If the job is cancelled by the daemon, this method won't be called.  Instead `node_failed_job` will be called.

Reply:
I would imagine that it is the same situation with timing out.  `node_completed_job` is only called for successfully completed jobs, but that job could be in the `CANCELLING` or `TIMINGOUT` state.

Reply:
So whether `node_completed_job` or `node_failed_job` is trigger depends on if `NODE_COMPLETED_JOB` or `NODE_FAILED_JOB` is sent from the daemon.

The cancellation of a job will trigger the daemon to send `SIGTERM` to the `batchd` or `stepd` process which in turn will send one of the above messages. Most processes *should* exit 143 after receiving the signal and in turn send `NODE_FAILED_JOB`. However they could equally likely say:

```
function cleanup {
  echo Time to confuse the controller, hehehe
  exit 0
}
trap cleanup SIGTERM 
```

In which case the `batchd` process will send `NODE_COMPLETED_JOB` even though the job has been cancelled.

Reply:
Interesting use case.  I think its probably better to handle the race condition I mentioned rather than scripts exiting `0` on receiving a `SIGTERM`.  Ideally, we'd be able to handle both, but that looks like a lot of effort.